src/rate_limiter.rs

```diff
@@ -181,89 +277,276 @@ impl RateLimiter {
         FakeClock::advance_time(wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client);
+        assert!(huge_message_can_be_added(&mut rate_limiter, &client));
+    }
+
+    #[test]
+    fn overcharge_correction() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+
+        let approx_num_parts = 10;
+        let data_size = approx_num_parts * MAX_PART_LEN;
+        let data = ImmutableData::new(vec![0; data_size]);
+        let msg_id = MessageId::new();
+
+        let request = UserMessage::Request(Request::GetIData {
+            name: *data.name(),
+            msg_id,
+        });
+        let request_parts = unwrap!(request.to_parts(0));
+
+        let charge = add_user_msg_part(&mut rate_limiter, &client, &request_parts[0]);
+        assert_eq!(unwrap!(charge), MAX_IMMUTABLE_DATA_SIZE_IN_BYTES);
+
+        let response = UserMessage::Response(Response::GetIData {
+            res: Ok(data),
+            msg_id,
+        });
+        let response_parts = unwrap!(response.to_parts(0));
+
+        let mut single_deduction = None;
+
+        for part in &response_parts {
+            if let Some(deduction) = discount_user_msg_part(&mut rate_limiter, &client, part) {
+                if single_deduction.is_none() {
+                    single_deduction = Some(deduction);
+                } else {
+                    panic!("deduction was applied more than once!");
+                }
+            }
+        }
+        let expected_deduction = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES -
+            (response_parts.len() * MAX_PART_LEN) as u64;
+        assert_eq!(single_deduction, Some(expected_deduction));
+    }
+
+    #[test]
+    fn prevent_msg_id_reuse_attack() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+
+        // Message ID used by both the put and the get.
+        let msg_id = MessageId::new();
+        let put_data = ImmutableData::new(vec![0; 4]);
+        let get_data = ImmutableData::new(vec![1; 10 * MAX_PART_LEN]);
+
+        let put_request = UserMessage::Request(Request::PutIData {
+            data: put_data,
+            msg_id,
+        });
+        let put_request_parts = unwrap!(put_request.to_parts(0));
+        let put_response = UserMessage::Response(Response::PutIData {
+            res: Ok(()),
+            msg_id,
+        });
+        let put_response_parts = unwrap!(put_response.to_parts(0));
+        let get_request = UserMessage::Request(Request::GetIData {
+            name: *get_data.name(),
+            msg_id,
+        });
+        let get_request_parts = unwrap!(get_request.to_parts(0));
+        let get_response = UserMessage::Response(Response::GetIData {
+            res: Ok(get_data),
+            msg_id,
+        });
+        let get_response_parts = unwrap!(get_response.to_parts(0));
+
+        // Put request hits the rate limiter first.
```

Comment:
what will happen if the get request arrives before the put?
looks like the client will ban (return with InvalidMessage).
shall it be confirmed in the test?